### DISCUSSION TITLE
Hackroms support

### DISCUSSION THREAD

**Discussion by XargonWan** [rommapp/romm#296] [Ideas]
Here is a possible hackrom support to be added to RomM.

**Features**
- Hackrom naming convention
- Ability to recognize if a rom is an hackrom
- IPS file storage
- IPS patcher
- Search for all the hacks applicable to a specific rom
- (opt) IPS shared "store" (maybe to discuss in another issue)
- Different saves/states for each IPS version management 
- The latest save/state set is available even for the next IPS version
- Storage side there will be saves with each IPS version name
- There may be added additional database entries such as:
  - `hack_version`
 
Example:
- Base rom: `MonsterBall Red (J).gba`
- `MonsterBall Red (J)[English translation v1.0].ips`
- `MonsterBall Red (J)[English translation v1.0].srm`
- `MonsterBall Red (J)[English translation v0.9].srm`

List example:
Base rom: `MonsterBall Red (J).gba`
IPSs:
`MonsterBall Red (J)[English translation]` <- this is a game entry
  - `MonsterBall Red (J)[English translation v1.0].srm`
  - `MonsterBall Red (J)[English translation v0.9].srm`
  
`MonsterBall Red (J)[Randomizer]` <- this is a game entry
  - `MonsterBall Red (J)[Randomizer v1.3].srm`
  - `MonsterBall Red (J)[Randomizer v1.1].srm`
  
`MonsterBall Red (J)[Total conversion mod]` <- this is a game entry
  - `MonsterBall Red (J)[Total conversion mod v0.1].srm`
  
**Saves:**
When I patch `MonsterBall Red (J).gba` to `MonsterBall Red (J)[English translation v1.0].gba` the `MonsterBall Red (J)[English translation v0.9].srm` is copied into `MonsterBall Red (J)[English translation v1.0].srm`.
Base rom: `MonsterBall Red (J).gba`
IPSs:
`MonsterBall Red (J)[English translation]` <- this is a game entry
  - `MonsterBall Red (J)[English translation v1.0].srm`
  - `MonsterBall Red (J)[English translation v0.9].srm`
  

**How the patching should work:**
- Start from a base rom
- Upload an IPS
- If a similar IPS filename is already uploaded, it will be listed as a newer/different version of the IPS
- If not, a IPS new game entry is created and on the base game a reference of the available hacks will be added.
- If unsure, or if wished by the user, the IPS can be moved in/out/to a new IPS entry.
- An IPS (or a list of IPS with similar names) is treated as a single rom with the difference that the base rom is indicated. As a full game it can be scraped. 
- When the patched game is requested (downloaded/synced) it will be patched on the go to avoid to keep a bloat of old versions on the server. Or only the latest version can be kept and the previous one can be generated on the go. This is valid only for those roms where the patching is quick (such as gba).

**Comment by XargonWan** [comment]
In addition to this, every game during the scraping can be matched with hackroms database and link or show the possible tools/translations/hacks, such as: https://www.romhacking.net/games/1656/

Of course the same support can be added for other websites if present.

**Comment by zurdi15** [comment]
Ohg good one, having a new hackrom section listing all the possible hackroms would be really nice, a lot of people will find excellent hackroms.
